### ISSUE TITLE
Channel Token Sign Generation

### ISSUE THREAD

**Issue by gauravtiwari** [centrifugal/rubycent#1]
Hey @FZambia 

I will be adding some changes for ruby library to this gem later today to generate channel and api sign token for authentication however, as for now I am getting "Permission Denied" when authenticating private channels. 

Here are the methods to authenticate private channels and generate token

``` ruby
# Auth method to authenticate private channels
  def auth
    if current_user
      data = {}
      sign = generate_channel_sign(
          params[:client], params[:channels], "{}"
      )
      data[channel] = {
          "sign": sign,
          "info": "{}"
      }
      render :json => data
    else
      render :text => "Not authorized", :status => '403'
    end
  end
```

``` ruby
  # Generates channel sign based on client, channel and info
  def generate_channel_sign(client, channel, info="")
    digest = OpenSSL::Digest::SHA256.new
    data =  "#{client}#{channel}#{info}"
    OpenSSL::HMAC.hexdigest(digest, $client.secret, data)
  end

# $client is instance variable for Centrifuge initialiser
```

Is I am doing anything wrong?

![screen shot 2015-08-18 at 13 26 03](https://cloud.githubusercontent.com/assets/771039/9329845/e51983b8-45ac-11e5-8500-730c86d53ae3.png)

Thanks

**Comment by FZambia** [comment]
Hello, could you show `subscribe` command `params` which were sent to Centrifugo and resulted in `permission denied` response?

Could you also show json `data` you return for signed in user?

**Comment by gauravtiwari** [comment]
Sure @FZambia 

Here is signed response: 

``` javascript
{"$notifications#usertaNJCj_yxsvMPPMu8WmP":{"sign":"9dce07fb9af437fff55e5efbcf84b3ea7ad4beebcc31ca8fef9d2e5c1d276ca2","info":"{}"}}

"project_secret": "b90f9cd7-c3b6-4404-7703-4c933183fb92"
```

Parameters sent: 

``` ruby
{"client"=>"c3613785-3b74-43d3-6989-67ade91ce048", "channels"=>"$notifications#usertaNJCj_yxsvMPPMu8WmP"}
```

**Comment by FZambia** [comment]
Do you know about user only special symbol `#` (see User Channel Boundary [here](http://fzambia.gitbooks.io/centrifugal/content/server/channels.html))? Is your user ID is `usertaNJCj_yxsvMPPMu8WmP`?

**Comment by arrowcircle** [comment]
Info should be "". AFAIK, latest Centrifugo uses "" empty string as default value. What version of gem and Cengtifuge you use?

**Comment by gauravtiwari** [comment]
@FZambia yes saw the user boundary, is there anything else that I should be beware when using user boundary compared to private channels? What's the difference between them?

**Comment by gauravtiwari** [comment]
@arrowcircle Latest version of gem and Centrifugo latest release - 0.2.3.

**Comment by FZambia** [comment]
I think Centrifugo returns `permission denied` because you use `#` in channel name. In this case only user with ID `usertaNJCj_yxsvMPPMu8WmP` can subscribe on this channel. This is just a feature that allows to restrict subscription on channel to specific user. I think your current user doesn't have ID `usertaNJCj_yxsvMPPMu8WmP`, right?

**Comment by gauravtiwari** [comment]
@FZambia  Yes - you are right, I actually got confused with documentation it uses user#id

Shall I just use ID without user prefix?

**Comment by FZambia** [comment]
Depends on what you want - If you want to use this limiting feature then using correct user ID after `#` without any prefixes should make things work. Or just don't use `#` in channel name. Please, try it, does it work?

**Comment by gauravtiwari** [comment]
@FZambia working now (it was due to incorrect user id). May be we can improve the documentation a bit without user prefix.

**Comment by FZambia** [comment]
Yep, it needs some clarification. Could you send pull request to this repo with `generate_channel_sign` method?

**Comment by gauravtiwari** [comment]
thanks for the help @FZambia I will add some methods related to channel and api sign later today to this gem.

**Comment by FZambia** [comment]
thanks!

**Comment by FZambia** [comment]
Also please note that `channels` in request params can actually be a list of private channels - see tornado code https://github.com/centrifugal/examples/blob/master/tornado_application/main.py#L105

I think it works for you due to how Ruby request params work as you have only single channel to subscribe

**Comment by gauravtiwari** [comment]
Yup I actually implemented that initially, but due to error I just narrowed it to down for debugging purposes. One thing, I noticed when using batch requests that, the request blocks the page load for a fraction of second - it like page is halted for sometime and then loads fully. Is it because of batch request is being processed? I will investigate into this too and see it happens again after the fix.
